File: src/main/java/com/github/tuk-cps/ExprParser/ExprTreeITE.java
package ExprParser;

import jAADD.BDD;
import static jAADD.AADD.REAL;

/**
 * The ITE function.
 */
class ExprTreeITE extends ExprTreeFunction {

    ExprTreeITE()  throws ExprError {
        super(REAL, "ite", new ExprTree("i", BDD.BOOL), new ExprTree("t", REAL), new ExprTree("e", REAL));
    }

    // ITE function
    @Override
    void evalThis() throws ExprError {
        assert param.size() == 3: "3 parameters expected";
        if (getParam(1).getResult() instanceof BDD)
            result = getBDDParam(0).ITE(getBDDParam(1), getBDDParam(2));
        else
            result = getBDDParam(0).ITE(getAADDParam(1), getAADDParam(2));
    }

    // ITE parameters with lazy evaluation.
    @Override
    void evalExpr() throws ExprError {
        getSymTab().scopes.push(this);

        getParam(0).evalExpr(); // Evaluate i
        BDD i=getBDDParam(0);   //
        if (i == BDD.ONE ) { getParam(1).evalExpr();}
        else if (i == BDD.ZERO) { getParam(2).evalExpr(); }
        else {
            getParam(1).evalExpr();
            getParam(2).evalExpr();
        }

        // now do ITE function.
        evalThis();

        getSymTab().scopes.pop();
    }
}

File: src/main/kotlin/com/github/tuk-cps/exprParser/ExprTreeITE.kt
package exprParser

import jAADD.AADD
import jAADD.BDD
import jAADD.BDD.Companion.variable

/**
 * The ITE function.
 */
internal class ExprTreeITE : ExprTreeFunction(AADD.Reals, "ite", ExprTree(variable("x"), "i"), ExprTree(AADD.Reals, "t"), ExprTree(AADD.Reals, "e")) {
    // ITE function
    @Throws(ExprError::class)
    override fun evalUp() {
        require(param.size == 3) { "3 parameters expected" }
        value = if (getParam(1).value is BDD) getBDDParam(0).ite(getBDDParam(1), getBDDParam(2))
        else getBDDParam(0).ite(getAADDParam(1), getAADDParam(2))
    }

    // ITE parameters with lazy evaluation.
    @Throws(ExprError::class)
    override fun evalUpRec() {
        symTab.scopes.push(this)
        getParam(0).evalUpRec() // Evaluate i
        val i = getBDDParam(0) //
        if (i === BDD.True) {
            getParam(1).evalUpRec()
        } else if (i === BDD.False) {
            getParam(2).evalUpRec()
        } else {
            getParam(1).evalUpRec()
            getParam(2).evalUpRec()
        }
        // now do ITE function.
        evalUp()
        symTab.scopes.pop()
    }
}